client/src/main/java/io/pravega/client/tables/TableSegment.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import io.pravega.common.util.AsyncIterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import lombok.Data;
+
+/**
+ * Defines all operations that are supported on a Table Segment.
+ *
+ * @param <KeyT>   Table Key Type.
+ * @param <ValueT> Table Value Type.
+ */
+public interface TableSegment<KeyT, ValueT> extends AutoCloseable {
+    /**
+     * Inserts or updates an existing Table Entry into this Table Segment.
+     * @param entry The Entry to insert or update. If {@link VersionedEntry#getVersion()} is null, this will perform a
+     *              Blind Update, otherwise it will perform a Conditional Update based on the information provided.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> put(VersionedEntry<KeyT, ValueT> entry);
+
+    /**
+     * Inserts new or updates existing Table Entries into this Table Segment.
+     *
+     * @param entries A list of entries to insert or update. If for at least one such entry, {@link VersionedKey#getVersion()}
+     *                returns a non-null value, this will perform an atomic Conditional Update where all the entries either
+     *                get applied or none will; otherwise a Blind Update will be performed.
+     * @return A CompletableFuture that, when completed, will contain a list of {@link KeyVersion}s associated with the
+     * newly inserted or updated entries. These will be in the same order as the entries given.
+     */
+    CompletableFuture<List<KeyVersion>> put(List<VersionedEntry<KeyT, ValueT>> entries);
```

Comment:
Since this is a batch update, we should have a way to know which keys failed the conditional check. Otherwise it may force the client to fetch all the keys and attempt an update.

Reply:
@shiveshr  That's a good point. I think we'll need to declare a new type of exception which contains this. I'll need to do this both server side and on the client.

Reply:
@tkaitchuck Fixed.